pkg/controller/clusterpool/collections.go

```diff
@@ -584,6 +599,252 @@ func (cds *cdCollection) Delete(c client.Client, cdName string) error {
+// ClusterDeploymentCustomizations that are related to specified pool.
+func getAllCustomizationsForPool(c client.Client, pool *hivev1.ClusterPool, logger log.FieldLogger) (*cdcCollection, error) {
+	if pool.Spec.Inventory == nil {
+		return nil, nil
+	}
+	cdcList := &hivev1.ClusterDeploymentCustomizationList{}
+	if err := c.List(
+		context.Background(), cdcList,
+		client.MatchingFields{claimClusterPoolIndex: pool.Name},
+		client.InNamespace(pool.Namespace)); err != nil {
+		logger.WithError(err).Error("error listing ClusterDeploymentCustomizations")
+		return nil, err
+	}
+
+	cdcCol := cdcCollection{
+		unassigned: make([]*hivev1.ClusterDeploymentCustomization, 0),
+		missing:    make([]string, 0),
+		reserved:   make([]*hivev1.ClusterDeploymentCustomization, 0),
+		cloud:      make([]*hivev1.ClusterDeploymentCustomization, 0),
+		syntax:     make([]*hivev1.ClusterDeploymentCustomization, 0),
+		byCDCName:  make(map[string]*hivev1.ClusterDeploymentCustomization),
+	}
+
+	for _, item := range pool.Spec.Inventory {
+		missing := true
+		for i, cdc := range cdcList.Items {
+			ref := &cdcList.Items[i]
+			if cdc.Name != item.Name {
+				continue
+			}
+			missing = false
+			cdcCol.byCDCName[item.Name] = ref
+			if cdRef := cdc.Status.ClusterDeploymentRef; cdRef == nil {
+				cdcCol.unassigned = append(cdcCol.unassigned, ref)
+			} else {
+				cdcCol.reserved = append(cdcCol.reserved, ref)
+			}
+			if cdc.Status.LastApplyStatus == hivev1.LastApplyBrokenCloud {
+				cdcCol.cloud = append(cdcCol.cloud, ref)
+			}
+			if cdc.Status.LastApplyStatus == hivev1.LastApplyBrokenSyntax {
+				cdcCol.syntax = append(cdcCol.cloud, ref)
+			}
+		}
+		if missing {
+			cdcCol.missing = append(cdcCol.missing, item.Name)
+		}
+	}
+
+	logger.WithFields(log.Fields{
+		"reservedCount":       len(cdcCol.reserved),
+		"unassignedCount":     len(cdcCol.unassigned),
+		"brokenByCloudCount":  len(cdcCol.cloud),
+		"brokenBySyntaxCount": len(cdcCol.syntax),
+		"missingCount":        len(cdcCol.missing),
+	}).Debug("found ClusterDeploymentCustomizations for ClusterPool")
+
+	return &cdcCol, nil
+}
+
+// SyncClusterDeploymentCustomizations makes sure that CDCs relagted to the pool, the pool inventory and related ClusterDeployments
+// are in the correct state.
+func (cdcs *cdcCollection) SyncClusterDeploymentCustomizationAssignments(c client.Client, pool *hivev1.ClusterPool, cds *cdCollection, logger log.FieldLogger) error {
+	if pool.Spec.Inventory == nil {
+		return nil
+	}
+
+	contains := func(s []string, str string) bool {
+		for _, v := range s {
+			if v == str {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	cdNames := []string{}
+	for cdName := range cds.byCDName {
+		cdNames = append(cdNames, cdName)
+	}
+
+	// If there is no CD, but CDC is reserved, then we release the CDC
+	for _, cdc := range cdcs.reserved {
+		if !contains(cdNames, cdc.Status.ClusterDeploymentRef.Name) {
```

Comment:
Again, I thought I already mentioned this, but...

We don't have to contsruct a `cdNames` list and spin through it (O(N)) to look up CD names. We can just use the `cds.ByName()` func.
```suggestion
		if cds.ByName(cdc.Status.ClusterDeploymentRef.Name) == nil {
```
Similar throughout.